src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs

```diff
@@ -0,0 +1,157 @@
+                        ? word.ToLowerInvariant(destination)
+                        : word.ToUpperInvariant(destination);
+
+                    if (written > 0)
+                    {
+                        break;
+                    }
+
+                    ExpandBuffer(ref result);
+                }
+
+                if (resultLength != 0)
+                {
+                    result[resultLength] = _separator;
+                    resultLength += 1;
+                }
+
+                resultLength += written;
+            }
+
+            int first = 0;
+            ReadOnlySpan<char> chars = name.AsSpan();
+            CharCategory previousCategory = CharCategory.Boundary;
+
+            for (int index = 0; index < chars.Length; index++)
+            {
+                char current = chars[index];
+                UnicodeCategory currentCategoryUnicode = char.GetUnicodeCategory(current);
+
+                if (currentCategoryUnicode == UnicodeCategory.SpaceSeparator ||
+                    currentCategoryUnicode >= UnicodeCategory.ConnectorPunctuation &&
+                    currentCategoryUnicode <= UnicodeCategory.OtherPunctuation)
+                {
+                    WriteWord(chars.Slice(first, index - first), ref result);
+
+                    previousCategory = CharCategory.Boundary;
+                    first = index + 1;
+
+                    continue;
+                }
+
+                if (index + 1 < chars.Length)
+                {
+                    char next = chars[index + 1];
+                    CharCategory currentCategory = currentCategoryUnicode switch
+                    {
+                        UnicodeCategory.LowercaseLetter => CharCategory.Lowercase,
+                        UnicodeCategory.UppercaseLetter => CharCategory.Uppercase,
+                        _ => previousCategory
+                    };
+
+                    if (currentCategory == CharCategory.Lowercase && char.IsUpper(next) ||
+                        next == '_')
+                    {
+                        WriteWord(chars.Slice(first, index - first + 1), ref result);
+
+                        previousCategory = CharCategory.Boundary;
+                        first = index + 1;
+
+                        continue;
+                    }
+
+                    if (previousCategory == CharCategory.Uppercase &&
+                        currentCategoryUnicode == UnicodeCategory.UppercaseLetter &&
+                        char.IsLower(next))
+                    {
+                        WriteWord(chars.Slice(first, index - first), ref result);
+
+                        previousCategory = CharCategory.Boundary;
+                        first = index;
+
+                        continue;
+                    }
+
+                    previousCategory = currentCategory;
+                }
+            }
+
+            WriteWord(chars.Slice(first), ref result);
+
+            name = result.Slice(0, resultLength).ToString();
```

Comment:
I had to double check if `ReadOnlySpan<char>.ToString()` is special cased so that it converts the buffer contents into a string (it appears to behave differently compared to ROS<byte> or ROS<byte>). Perhaps make the transformation more explicit using the `new string(ROS<char>)` constructor?

Reply:
For any span over characters `ToString()` behaves differently and does exactly the same as you wrote. Otherwise, [a published prototype on NuGet](https://www.nuget.org/packages/Yoh.Text.Json.NamingPolicies) for almost one and a half years wouldn't work (:

Reply:
I mean, it's obvious that it works (otherwise your tests would be failing :-)). I'm just approaching it from the point of view that myself and probably others have been trained to interpret the output of this as something like "System.ReadOnlySpan<Char>[x]"`

Reply:
Unfortunately, that part as others like `Equals(object)` and `GetHashCode()` came from .NET Fx 1 which had no generics (as Jeffry Richter said, the team had two options at that time: release without generics which already were in work or being fired; they decided that the first option looks better). In Rust that issue is solved by traits which has constraints in generic cases, so if that would be possible in .NET then `ToString` can be available only for types which support conversion to string.